lib/spack/spack/installer.py

```diff
@@ -365,12 +366,13 @@ def _update_explicit_entry_in_db(pkg, rec, explicit):
             ``False`` if it was pulled in as a dependency of an explicit
             package.
     """
-    if explicit and not rec.explicit:
+    if explicit != rec.explicit:
         with spack.store.db.write_transaction():
             rec = spack.store.db.get_record(pkg.spec)
-            message = '{s.name}@{s.version} : marking the package explicit'
-            tty.msg(message.format(s=pkg.spec))
-            rec.explicit = True
+            message = '{s.name}@{s.version} : marking the package {0}'
+            status = 'explicit' if explicit else 'implicit'
+            tty.msg(message.format(status, s=pkg.spec))
+            rec.explicit = explicit
```

Comment:
I think this functionality should move to `lib/spack/spack/database.py`.

Reply:
I have now made it a proper method of the Database class.